### PR TITLE
[Python] Add the ability to provide a list of files to `read_csv` 

### LINKED ISSUE #8822: [python] read_csv should accept iterable of filenames
### What happens?

In the Python api, passing a list of strings of filename to `read_csv` errors.

In the CLI, this works fine.  Probably just a matter of wiring up the python bindings a little bit differently.

### To Reproduce

```
D select count(*) from read_csv_auto('ci/ibis-testing-data/csv/batting.csv');
┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│       101332 │
└──────────────┘
D select count(*) from read_csv_auto(['ci/ibis-testing-data/csv/batting.csv', 'ci/ibis-testing-data/csv/batting.csv']);
┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│       202664 │
└──────────────┘
```

vs

```python

[ins] In [1]: import duckdb

[ins] In [2]: con = duckdb.connect()

[ins] In [3]: con.read_csv(
         ...:     ["ci/ibis-testing-data/csv/batting.csv", "ci/ibis-testing-data/csv/batting.csv"]
         ...: )
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[3], line 1
----> 1 con.read_csv(
      2     ["ci/ibis-testing-data/csv/batting.csv", "ci/ibis-testing-data/csv/batting.csv"]
      3 )

File /nix/store/pl3rxj6p2s8mx25z469vs6ccr4xvm36f-python3-3.10.12-env/lib/python3.10/site-packages/pyduckdb/filesystem.py:56, in ModifiedMemoryFileSystem.add_file(self, object, path)
     54 def add_file(self, object, path):
     55 	if not is_file_like(object):
---> 56 		raise ValueError("Can not read from a non file-like object")
     57 	path = self._strip_protocol(path)
     58 	if isinstance(object, TextIOBase):
     59 		# Wrap this so that we can return a bytes object from 'read'

ValueError: Can not read from a non file-like object
```

### OS:

ubuntu 22.04.2 x86

### DuckDB Version:

v0.8.1

### DuckDB Client:

Python and CLI

### Full Name:

Gil Forsyth

### Affiliation:

Voltron Data

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by Tishj**
This PR fixes #8822 

We now accept things like:
```py
import duckdb
from io import StringIO

con = duckdb.connect()
file1 = StringIO('one,two,three,four\n1,2,3,4\n1,2,3,4\n1,2,3,4')
file2 = StringIO('one,two,three,four\n5,6,7,8\n5,6,7,8\n5,6,7,8')
file3 = StringIO('one,two,three,four\n9,10,11,12\n9,10,11,12\n9,10,11,12')
files = [file1, file2, file3]
rel = con.read_csv(files)
res = rel.fetchall()
assert res == [
	(1, 2, 3, 4),
	(1, 2, 3, 4),
	(1, 2, 3, 4),
	(5, 6, 7, 8),
	(5, 6, 7, 8),
	(5, 6, 7, 8),
	(9, 10, 11, 12),
	(9, 10, 11, 12),
	(9, 10, 11, 12),
]
```

### CODE REVIEW

File: tools/pythonpkg/tests/fast/api/test_read_csv.py [hunk 1]
```diff
@@ -544,3 +544,23 @@ def test_read_csv_names_mixed_with_dtypes(self):
...
                 },
             )
+
+    def test_read_csv_multi_file(self):
+        con = duckdb.connect()
+        file1 = StringIO('one,two,three,four\n1,2,3,4\n1,2,3,4\n1,2,3,4')
+        file2 = StringIO('one,two,three,four\n5,6,7,8\n5,6,7,8\n5,6,7,8')
+        file3 = StringIO('one,two,three,four\n9,10,11,12\n9,10,11,12\n9,10,11,12')
+        files = [file1, file2, file3]
+        rel = con.read_csv(files)
```
Comment (Mytherin) [inline tools/pythonpkg/tests/fast/api/test_read_csv.py:554]: Can you test this with an empty list?

File: tools/pythonpkg/tests/fast/api/test_read_csv.py [hunk 2]
```diff
@@ -544,3 +544,23 @@ def test_read_csv_names_mixed_with_dtypes(self):
...
+        res = rel.fetchall()
```
Comment (Mytherin) [inline tools/pythonpkg/tests/fast/api/test_read_csv.py:555]: Can you also test this with a file that doesn't exist in the middle of the list?

File: tools/pythonpkg/src/include/duckdb_python/filesystem_object.hpp [hunk 3]
```diff
@@ -14,18 +14,23 @@ namespace duckdb {
...
-		obj.attr("delete")(filename);
+		auto &files = ListValue::GetChildren(filenames);
+		for (auto &file : files) {
+			auto name = StringValue::Get(file);
+			obj.attr("delete")(name);
+		}
 	}
 
-	string filename;
+	Value filenames;
```
Comment (Mytherin) [inline tools/pythonpkg/src/include/duckdb_python/filesystem_object.hpp:33]: Maybe this should be a `vector<string>` instead?

File: src/main/relation/read_csv_relation.cpp [hunk 4]
```diff
@@ -36,16 +49,13 @@ ReadCSVRelation::ReadCSVRelation(const shared_ptr<ClientContext> &context, const
 	AddNamedParameter("columns", Value::STRUCT(std::move(column_names)));
 }
 
-ReadCSVRelation::ReadCSVRelation(const std::shared_ptr<ClientContext> &context, const string &csv_file,
+ReadCSVRelation::ReadCSVRelation(const std::shared_ptr<ClientContext> &context, const Value &input,
```
Comment (Mytherin) [inline src/main/relation/read_csv_relation.cpp:52]: Maybe have two separate constructors - one with a `string` and one with `vector<string>`? The single string can just call the `vector<string>` with a list.

File: src/main/connection.cpp [hunk 5]
```diff
@@ -223,8 +223,8 @@ shared_ptr<Relation> Connection::ReadCSV(const string &csv_file) {
 	return ReadCSV(csv_file, std::move(options));
 }
 
-shared_ptr<Relation> Connection::ReadCSV(const string &csv_file, named_parameter_map_t &&options) {
-	return make_shared<ReadCSVRelation>(context, csv_file, std::move(options));
+shared_ptr<Relation> Connection::ReadCSV(const Value &csv_input, named_parameter_map_t &&options) {
```
Comment (Mytherin) [inline src/main/connection.cpp:226]: Same here - can we have two constructors with `string` and `vector<string>`

File: tools/pythonpkg/src/include/duckdb_python/path_like.hpp [hunk 6]
```diff
@@ -3,14 +3,16 @@
...
 
 namespace duckdb {
 
 struct DuckDBPyConnection;
 
 struct PathLike {
 	static PathLike Create(const py::object &object, DuckDBPyConnection &connection);
-	string str;
+	// The file(s) extracted from object
+	Value files;
```
Comment (Mytherin) [inline tools/pythonpkg/src/include/duckdb_python/path_like.hpp:15]: `vector<string>`

File: tools/pythonpkg/src/path_like.cpp [hunk 7]
```diff
@@ -4,26 +4,90 @@
...
+		}
+		return *object_store;
+	}
+
+public:
+	DuckDBPyConnection &connection;
+	optional_ptr<ModifiedMemoryFileSystem> object_store;
+	PythonImportCache &import_cache;
+	// The list containing every file
+	vector<Value> all_files;
```
Comment (Mytherin) [inline tools/pythonpkg/src/path_like.cpp:34]: Same here - `vector<string>`

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good - some minor comments:

**Comment by Mytherin** [comment]
Looks like the CI is actually broken here - could you have a look?

**Comment by FrancoisLepoutre** [comment]
Hi all,

I was afraid to ask for StringIO-based input within the python stuff since I already asked for so much that I got it for free (no coding!). 

Glad to hear you will be able to feed the beast from in-memory stuff as well. Yes, that helps a lot as there are quite a number of decent python-based use cases, including feeding from arbitrary structured text formats that are not csv or parquet based.

Great news:-)

Cheers

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**